### PR TITLE
LPS-166083 Fixed creating duplicate TestFlow

### PR DESCRIPTION
**Pull request by rafaelfs98**

### CODE REVIEW

**Comment by rafaelfs98** [comment]
ci:test:sf

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/TestflowModal.tsx [hunk 1]
```diff
@@ -38,10 +40,19 @@ const TestflowModal: React.FC<TestflowModalProps> = ({
...
+		const taskResponse = await testrayTaskImpl.getTasksByBuildId(buildId);
+
+		if (taskResponse?.totalCount) {
+			return Liferay.Util.openToast({
+				message: i18n.translate('a-task-for-this-build-already-exists'),
+				type: 'danger',
+			});
 		}
+
+		return navigate(`/testflow/${buildId}/create`);
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/TestflowModal.tsx:55]: No need to return

```suggestion
	navigate(`/testflow/${buildId}/create`);
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCase.ts [hunk 2]
```diff
@@ -61,6 +63,34 @@ class TestrayCaseRest extends Rest<Case, TestrayCase> {
 			uri: 'cases',
 		});
 	}
+	protected async validate(Case: Case, id?: number) {
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCase.ts:66]: Extra space

```suggestion

	protected async validate(Case: Case, id?: number) {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCase.ts [hunk 3]
```diff
@@ -61,6 +63,34 @@ class TestrayCaseRest extends Rest<Case, TestrayCase> {
...
+
+		const response = await this.fetcher<APIResponse<TestrayCase>>(
+			`/cases?filter=${filters}`
+		);
+
+		if (response?.totalCount) {
+			throw new Error(i18n.sub('the-x-name-already-exists', 'case'));
+		}
+	}
+	protected async beforeCreate(Case: Case): Promise<void> {
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCase.ts:87]: Space

```suggestion

	protected async beforeCreate(Case: Case): Promise<void> {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts [hunk 4]
```diff
@@ -12,19 +12,47 @@
...
+class TestrayCaseTypeImpl extends Rest<CaseType, TestrayCaseType> {
+	constructor() {
+		super({
+			adapter: ({name}) => ({
+				name,
+			}),
+			uri: 'casetypes',
+		});
+	}
+	protected async validate(suite: CaseType, id?: number) {
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts:32]:
```suggestion

	protected async validate(caseType: CaseType, id?: number) {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts [hunk 5]
```diff
@@ -12,19 +12,47 @@
...
-export {createCaseTypes, updateCaseTypes};
+		const response = await this.fetcher<APIResponse<TestrayCaseType>>(
+			`/casetypes?filter=${filters}`
+		);
+
+		if (response?.totalCount) {
+			throw new Error(i18n.sub('the-x-name-already-exists', 'case-type'));
+		}
+	}
+	protected async beforeCreate(suite: CaseType): Promise<void> {
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts:49]:
```suggestion

	protected async beforeCreate(caseType: CaseType): Promise<void> {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts [hunk 6]
```diff
@@ -12,19 +12,47 @@
...
 
-const createCaseTypes = (casetype: CaseType) =>
-	fetcher.post('/casetypes', adapter(casetype));
+		if (id) {
+			searchBuilder.ne('id', id).and();
+		}
 
-const updateCaseTypes = (id: number, casetype: CaseType) =>
-	fetcher.put(`/casetypes/${id}`, adapter(casetype));
+		const filters = searchBuilder.eq('name', suite.name).build();
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts:39]:
```suggestion
		const filters = searchBuilder.eq('name', caseType.name).build();
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts [hunk 7]
```diff
@@ -12,19 +12,47 @@
...
+		const response = await this.fetcher<APIResponse<TestrayCaseType>>(
+			`/casetypes?filter=${filters}`
+		);
+
+		if (response?.totalCount) {
+			throw new Error(i18n.sub('the-x-name-already-exists', 'case-type'));
+		}
+	}
+	protected async beforeCreate(suite: CaseType): Promise<void> {
+		await this.validate(suite);
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts:50]:
```suggestion
		await this.validate(caseType);
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts [hunk 8]
```diff
@@ -12,19 +12,47 @@
...
+	}
+
+	protected async beforeUpdate(id: number, suite: CaseType): Promise<void> {
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts:53]:
```suggestion
	protected async beforeUpdate(id: number, caseType: CaseType): Promise<void> {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts [hunk 9]
```diff
@@ -12,19 +12,47 @@
...
+		await this.validate(suite, id);
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayCaseTypes.ts:54]:
```suggestion
		await this.validate(caseType, id);
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayRoutine.ts [hunk 10]
```diff
@@ -32,6 +34,39 @@ class TestrayRoutineImpl extends Rest<RoutineFormType, TestrayRoutine> {
...
 	}
+
+	protected async validate(routine: RoutineFormType, id?: number) {
+		const searchBuilder = new SearchBuilder();
+
+		if (id) {
+			searchBuilder.ne('id', id).and();
+		}
+
+		const filters = searchBuilder
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayRoutine.ts:45]:
```suggestion
		const filter = searchBuilder
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayRoutine.ts [hunk 11]
```diff
@@ -32,6 +34,39 @@ class TestrayRoutineImpl extends Rest<RoutineFormType, TestrayRoutine> {
...
+			.eq('name', routine.name)
+			.and()
+			.eq('projectId', routine.projectId)
+			.build();
+
+		const response = await this.fetcher<APIResponse<TestrayRoutine>>(
+			`/routines?filter=${filters}`
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestrayRoutine.ts:52]:
```suggestion
			`/routines?filter=${filter}`
```

**Review by kevenleone — CHANGES_REQUESTED** [review]

**Comment by kevenleone** [comment]
ci:forward

**Comment by kevenleone** [comment]
ci:forward

**Review by welitonpe — APPROVED** [review]

**Pull request closed without merging**